### ISSUE TITLE
Epoch data not aligned when stick_to_shard=True

### ISSUE THREAD

**Issue by kiyoon** [NVIDIA/DALI#1762]
Hi, I'm trying to make a dataloader that gives you the same data every epoch for validation and testing.
I realised that if I set the pad_last_epoch=False, the next epoch data is always shifted and dropped.
So I set it always to True, and I am just changing the size of the DALIGenericIterator.

It seemed to work well for the first 2 epoch, but weirdly it gets shifted by one sample every epoch from the 3rd one. I spotted this when I set stick_to_shard=True, because it is hard to see the effect when it's False.
Is there anything I am missing or is it a bug from the VideoReader?

I have 922 samples, and since the number of shards is 3, each shard (with padding) is of size 308.

```python
from nvidia.dali.pipeline import Pipeline
from nvidia.dali.pipeline import Pipeline
from nvidia.dali.plugin import pytorch
import nvidia.dali.ops as ops
import nvidia.dali.types as types

import torch
import math
import numpy as np

from video_datasets_api.epic_kitchens.read_annotations import get_verb_uid3label_dict

class VideoReaderPipeline(Pipeline):
    def __init__(self, batch_size, sequence_length, stride, crop_size, num_threads, file_list, seed, step, device_id, shard_id, num_shards, shuffle, mean, std, pad_last_batch):
        super(VideoReaderPipeline, self).__init__(batch_size, num_threads, device_id, seed=seed)
        self.reader = ops.VideoReader(device="gpu", file_list=file_list, sequence_length=sequence_length, normalized=True,
                                    random_shuffle=shuffle, image_type=types.RGB, dtype=types.FLOAT, initial_fill=33, enable_frame_num=True, stride=stride, step=step, shard_id=shard_id, num_shards=num_shards, stick_to_shard=True, pad_last_batch=pad_last_batch)
        self.uniform = ops.Uniform(range=(1.0, 1.0))
        self.coin = ops.CoinFlip(probability=1.5)
        self.cropmirrornorm = ops.CropMirrorNormalize(device="gpu", crop=crop_size, output_dtype=types.FLOAT, mean = mean, std = std, output_layout = "CFHW")

    def define_graph(self):
        input = self.reader(name="Reader")
        crop_pos_x = self.uniform()
        crop_pos_y = self.uniform()
        is_flipped = self.coin()
        output = self.cropmirrornorm(input[1], crop_pos_x=crop_pos_x, crop_pos_y=crop_pos_y, mirror=is_flipped)
        # Change what you want from the dataloader.
        # input[2]: label, input[2]: starting frame number indexed from zero
        return output, input[2], input[2], crop_pos_x, crop_pos_y, is_flipped

class DALILoader():
    def __init__(self, batch_size, file_list, uid3label, sequence_length, stride, crop_size, seed, step=-1, num_threads=1, device_id=0, shard_id=0, num_shards=1, shuffle=False, mean=[0.45, 0.45, 0.45], std=[0.225, 0.225, 0.225], fill_last_batch=False, pad_last_batch=True):
        self.pipeline = VideoReaderPipeline(batch_size=batch_size,
                                            sequence_length=sequence_length,
                                            stride=stride,
                                            num_threads=num_threads,
                                            file_list=file_list,
                                            crop_size=crop_size,
                                            seed=seed,
                                            step=step,
                                            device_id=device_id,
                                            shard_id=shard_id,
                                            num_shards=num_shards,
                                            shuffle=shuffle,
                                            mean=mean,
                                            std=std,
                                            # It's always good to pad the last batch, otherwise some data from next epoch are dropped.
                                            # Instead, pad_last_batch changes the size of the shard (epoch)
                                            pad_last_batch=True)
        self.pipeline.build()

        self.batch_size = batch_size
        self.epoch_size = self.pipeline.epoch_size("Reader")

        if pad_last_batch:
            self.shard_size = math.ceil(self.epoch_size / num_shards)
        else:
            self.shard_size = int(self.epoch_size * (shard_id + 2) / num_shards) - int(self.epoch_size * shard_id / num_shards)

        self.num_iters = self.shard_size // batch_size + int(self.shard_size % batch_size > 1)
        self.dali_iterator = pytorch.DALIGenericIterator(self.pipeline,
                                                         ["data", "uid", "frame_num", "crop_pos_x", "crop_pos_y", "is_flipped"],
                                                         size=self.shard_size,
                                                         auto_reset=True,
                                                         fill_last_batch=fill_last_batch,
                                                         # It's always good to pad the last batch, otherwise some data from next epoch are dropped.
                                                         # Instead, pad_last_batch changes the size of the shard (epoch)
                                                         last_batch_padded=True)
    def __len__(self):
        return int(self.shard_size)
    def __iter__(self):
        #print(self.dali_iterator.__iter__())
        return self
    def __next__(self):
        batch = self.dali_iterator.__next__()[1]
        postprocessed = {}
        postprocessed['data'] = batch['data']
        # DALI uses the same buffer so you can't change the shape directly. You must copy them.
        postprocessed['uid'] = batch['uid'].view(-1)
        postprocessed['frame_num'] = batch['frame_num'].view(-1)
        postprocessed['crop_pos_x'] = batch['crop_pos_x'].view(-1)
        postprocessed['crop_pos_y'] = batch['crop_pos_y'].view(-1)
        postprocessed['is_flipped'] = batch['is_flipped'].view(-1)
        return postprocessed 


if __name__ == "__main__":
    import argparse
    parser = argparse.ArgumentParser()
    parser.add_argument('--file_list', type=str, default='file_list.txt',
                        help='DALI file_list for VideoReader')
    parser.add_argument('--frames', type=int, default = 16,
                        help='num frames in input sequence')
    parser.add_argument('--stride', type=int, default = 2,
                        help='temporal stride')
    parser.add_argument('--crop_size', type=int, nargs=2, default=[224, 224],
                        help='[height, width] for input crop')
    parser.add_argument('--batchsize', type=int, default=2,
                        help='per rank batch size')

    args = parser.parse_args()

    uid2label = get_verb_uid2label_dict('/disk/scratch1/s1884147/datasets/EPIC_KITCHENS_2018/annotations/EPIC_train_action_labels.pkl')
    loader = DALILoader(args.batchsize,
                args.file_list,
                uid2label,
                args.frames,
                args.stride,
                args.crop_size,
                shard_id=0,
                num_shards=3,
                seed=12,
                pad_last_batch=True
                )
    num_samples = len(loader)
    print("Total number of samples: %d" % num_samples)

    sample_seen = 0
    for batch in loader:
        sample_seen += batch['data'].shape[0]
        print("%d / %d" % (sample_seen, num_samples))
        print(batch["uid"])
        print(batch["frame_num"])

    sample_seen = 0
    for batch in loader:
        sample_seen += batch['data'].shape[0]
        print("%d / %d" % (sample_seen, num_samples))
        print(batch["uid"])
        print(batch["frame_num"])

    sample_seen = 0
    for batch in loader:
        sample_seen += batch['data'].shape[0]
        print("%d / %d" % (sample_seen, num_samples))
        print(batch["uid"])
        print(batch["frame_num"])

    sample_seen = 0
    for batch in loader:
        sample_seen += batch['data'].shape[0]
        print("%d / %d" % (sample_seen, num_samples))
        print(batch["uid"])
        print(batch["frame_num"])
```

**Comment by JanuszL** [comment]
Hi,

In 0.18 there was some bug fixed in #1557 that sounds like what you are experiencing:
```
- fixes problem with wrong cloning of the last sample in the batch when pad_last_batch is enabled and the reader needs to stick to the shard
```
Can you try the latest nightly/weekly build and see how it works?

**Comment by kiyoon** [comment]
For some reason I failed to install the nightly (in fact, any version) build. Is the extra index url link dead?

Also, what is the expected behaviour when the `stick_to_shard` is False?
Let's say you have 3 shards, and you have part1, part2, and part3 of the data.
`shard_id=0` gives you part1 for the first epoch, part2 for the second epoch, part3 for the third epoch, part1 again for the fourth epoch.
`shard_id=1` gives you part2, part3, part1, and part2 for epoch 1, 2, 3, 4.
`shard_id=2` gives you part3, part1, part2, part3.

As far as I understand, this is how it should work. Am I right about this?

**Comment by JanuszL** [comment]
`pip install --extra-index-url https://developer.download.nvidia.com/compute/redist/nightly/cuda/10.0 nvidia-dali-nightly` works fine for me.
If `stick_to_shard` is set to False and `pad_last_batch` to True as well, then it works as you described. If `pad_last_batch` is set to False and shard has the different size the GPU assigned to smaller one will start consuming next shard before the end of the epoch.

**Comment by kiyoon** [comment]
I see this error message when I try to install. I made a conda environment and installed PyTorch 1.4.0, and now I'm trying to install DALI.

```
ERROR: Could not find a version that satisfies the requirement nvidia-dali-nightly (from versions: none)
ERROR: No matching distribution found for nvidia-dali-nightly
```

Am I missing some dependencies or doing something wrong? I used to work with other environments.

**Comment by JanuszL** [comment]
What env you have (OS, architecture, python version) and how you have installed it so far?

**Comment by kiyoon** [comment]
It's Ubuntu 18.04 server, x86-64, python 3.8.1.

I did
```
conda create -n dali python=3
conda activate dali
conda install pytorch torchvision cudatoolkit=10.1 -c pytorch
pip install --extra-index-url https://developer.download.nvidia.com/compute/redist/nightly/cuda/10.0 nvidia-dali-nightly
```

**Comment by JanuszL** [comment]
Python 3.8 is not yet supported that is why you cannot install DALI. We will add 3.8 support soon, but I commit to a particular release yet.

**Comment by kiyoon** [comment]
Hi, I finally tested it with the nightly build and I see no problem. It should be already fixed then. Thank you so much!

**Comment by JanuszL** [comment]
0.19 is online, you can test with the regular release now.
